File: 2025/day02-2.hs
import Data.List (isPrefixOf)
import Distribution.Simple.Utils (safeTail)

type Range = (Int, Int)

solution :: String -> Int
solution = sum . concatMap (filter isInvalidId . unwrap . parseRange) . parse

unwrap :: Range -> [Int]
unwrap (start, end) = [start .. end]

isInvalidPattern :: String -> String -> Bool
isInvalidPattern _ "" = True
isInvalidPattern ps xs = ps `isPrefixOf` xs && isInvalidPattern ps (drop (length ps) xs)

isInvalidId :: Int -> Bool
isInvalidId n = any (\l -> isInvalidPattern (take l digits) digits) [1 .. half]
  where
    digits = show n
    half = length digits `div` 2

parseRange :: String -> Range
parseRange line = (read . takeWhile (/= '-') $ line, read . tail . dropWhile (/= '-') $ line)

parse :: String -> [String]
parse "" = []
parse content = takeWhile (/= ',') content : (parse . safeTail . dropWhile (/= ',') $ content)

main :: IO ()
main = do readFile "day02.dat" >>= print . solution

File: 2025/day02_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: str) -> Any:
    return sum(
        item
        for rng in (parse(block) for block in content.split(","))
        for item in rng
        if is_invalid(item)
    )


def is_invalid(item: int) -> bool:
    half = len(str(item)) // 2
    for n in range(half, 0, -1):
        start = str(item)[:n]
        if is_repeating(start, str(item)):
            return True
    return False


def is_repeating(start: str, rest: str) -> bool:
    if start == rest:
        return True
    if rest.startswith(start):
        return is_repeating(start, rest[len(start) :])
    return False


def parse(block: str) -> range:
    start, end = block.split("-")
    return range(int(start), int(end) + 1)


def main() -> None:
    with open("day02.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()
